staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit_test.go

```diff
@@ -855,82 +589,69 @@ func TestAuditJson(t *testing.T) {
+			sink := &fakeAuditSink{}
+			policyChecker := policy.FakeChecker(auditinternal.LevelRequestResponse)
+			handler := WithAudit(http.HandlerFunc(test.handler), &fakeRequestContextMapper{
+				user: &user.DefaultInfo{Name: "admin"},
+			}, sink, policyChecker, func(r *http.Request, ri *request.RequestInfo) bool {
+				// simplified long-running check
+				return ri.Verb == "watch"
+			})
+
+			req, _ := http.NewRequest(test.verb, test.path, nil)
+			if test.auditID != "" {
+				req.Header.Add("Audit-ID", test.auditID)
+			}
+			req.RemoteAddr = "127.0.0.1"
 
-		w := httptest.NewRecorder()
-		func() {
-			defer func() {
-				recover()
+			func() {
+				defer func() {
+					recover()
+				}()
+				handler.ServeHTTP(httptest.NewRecorder(), req)
 			}()
-			handler.ServeHTTP(w, req)
-		}()
 
-		t.Logf("[%s] audit log: %v", test.desc, buf.String())
+			events := sink.Events()
+			t.Logf("audit log: %#v", events)
 
-		line := strings.Split(strings.TrimSpace(buf.String()), "\n")
-		if len(line) != len(test.expected) {
-			t.Errorf("[%s] Unexpected amount of lines in audit log: %d", test.desc, len(line))
-			continue
-		}
-		expectedID := types.UID("")
-		for i, expect := range test.expected {
-			// decode events back to check json elements.
-			event := &auditv1beta1.Event{}
-			decoder := audit.Codecs.UniversalDecoder(auditv1beta1.SchemeGroupVersion)
-			if err := runtime.DecodeInto(decoder, []byte(line[i]), event); err != nil {
-				t.Errorf("failed decoding line %s: %v", line[i], err)
-				continue
-			}
-			if "admin" != event.User.Username {
-				t.Errorf("[%s] Unexpected username: %s", test.desc, event.User.Username)
+			if len(events) != len(test.expected) {
+				t.Fatalf("Unexpected amount of lines in audit log: %d", len(events))
 			}
-			if event.Stage != expect.Stage {
-				t.Errorf("[%s] Unexpected Stage: %s", test.desc, event.Stage)
+			expectedID := types.UID("")
+			for i, expect := range test.expected {
+				event := events[i]
+				if "admin" != event.User.Username {
+					t.Errorf("Unexpected username: %s", event.User.Username)
+				}
+				if event.Stage != expect.Stage {
+					t.Errorf("Unexpected Stage: %s", event.Stage)
+				}
+				if event.Verb != expect.Verb {
+					t.Errorf("Unexpected Verb: %s", event.Verb)
+				}
+				if event.RequestURI != expect.RequestURI {
+					t.Errorf("Unexpected RequestURI: %s", event.RequestURI)
+				}
+				if test.auditID != "" && event.AuditID != types.UID(test.auditID) {
+					t.Errorf("Unexpected AuditID in audit event, AuditID should be the same with Audit-ID http header")
+				}
+				if expectedID == types.UID("") {
+					expectedID = event.AuditID
+				} else if expectedID != event.AuditID {
+					t.Errorf("Audits for one request should share the same AuditID, %s differs from %s", expectedID, event.AuditID)
+				}
+				if (event.ResponseStatus == nil) != (expect.ResponseStatus == nil) {
+					t.Errorf("Unexpected ResponseStatus: %v", event.ResponseStatus)
+					continue
```

Comment:
> But this particular place has continue afterwards :)

Do you mean the `continue` in this line?
I think the `continue` here is OK.
For example, there are 3 events in test.expected totally. And it fails in line 646 for the first event, so we run `continue`. And the check in line 650 is skipped, and then we will check the next 2 events.